Update and edit advanced LDAP docs to use BasicAuthPasswordIdentityProvider

kalexand-rh opened the pull request:
continuation of https://github.com/openshift/openshift-docs/pull/9025

kalexand-rh commented:
We still need to figure out where to tell users this:
`This does have the negative side effect of making OpenShift see
user's passwords. However, any individual that is concerned by this
should instead use the RequestHeaderIdentityProvider with SAML and
GSSAPI to eliminate the need for users to type passwords at all.`

kalexand-rh commented:
I had a chat with Mo. We're going to just have one topic with only the option to configure LDAP and extend the email attribute.

kalexand-rh commented:
@enj, it's sliced and diced and reassembled into one topic. PTAL.

kalexand-rh commented:
@vikram-redhat, do you have a suggestion for QE, and do I need to set up redirects for the topics I removed?

vikram-redhat commented:
@kalexand-rh - as per the QE chart, it should be @stuartchuan for all auth components. Yes, we will need redirects. Can you add the redirects required to the register (not posting the link here - it should be available in our manual)?

kalexand-rh commented:
Thank you, Vikram. @stuartchuan, will you PTAL?

stuartchuan commented on install_config/topics/sssd_for_ldap_failover_prereqs.adoc:
```diff
@@ -0,0 +1,40 @@
+
+* Before starting setup, you need to know the following information about your
+LDAP server.
++
```
The following 4 lines need retract to make as subset.

stuartchuan commented on install_config/topics/sssd_for_ldap_failover_prereqs.adoc:
```diff
@@ -0,0 +1,40 @@
+* Whether the LDAP server corresponds to RFC 2307 or RFC2307bis for user groups.
++
+* Prepare the servers:
++
```
The following need retract to make as sbuset, refer https://docs.openshift.com/container-platform/3.4/install_config/advanced_ldap_configuration/sssd_for_ldap_failover.html#sssd-prerequisites-for-authenticating-proxy-setup

kalexand-rh commented:
Thanks @stuartchuan! I updated the nesting. Please let me know what else needs to change.

stuartchuan commented:
@kalexand-rh I have tried to verify the steps with real environment and setup up the sssd for LDAP failover following the new guide. When I go to the sssd_for_ldap_failover_configure_apache.adoc step, the apache server failed to setup, when I access https://remote-basic.example.com/check_user.php always get the
`Forbidden

You don't have permission to access /check_user.php on this server.`

@enj would you please help review this part, or give me some suggestion, then I could continue the setup.

enj commented:
@stuartchuan the issue likely lies in your SSSD or PAM configuration.  Check the Apache proxy logs which is configured in the docs as:

```
  # Send logs to a specific location to make them easier to find
  ErrorLog logs/remote_basic_error_log
  TransferLog logs/remote_basic_access_log
  LogLevel warn
```

Set `LogLevel` to something higher such as `debug` if `warn` does not indicate the issue.

kalexand-rh commented on install_config/configuring_authentication.adoc:
```diff
@@ -705,11 +715,17 @@ Basic Authentication is a generic backend integration mechanism that allows
+Because basic authentication is generic, you can use this identity
+provider for advanced authentication configurations, such as:
+* xref:../install_config/sssd_for_ldap_failover.adoc.adoc#setting-up-for-ldap-failover[LDAP failover]
+* link:https://github.com/openshift/request-header-saml-service-provider[SAML authentication]
```
Moving to line 948

enj replied:
Flow does not seem correct.  I would think it would be near `provide Kerberos, LDAP, and many other forms of enterprise authentication` or at the end (so right before `[[GitHub]]`).

kalexand-rh replied:
@enj, it doesn't really work anywhere, but if you think it's better up a paragraph, done. :)

enj commented on install_config/topics/sssd_for_ldap_failover_configure_apache.adoc:
```diff
@@ -0,0 +1,193 @@
+* install_config/sssd_for_ldap_failover.adoc
+////
+
+[id='sssd-configuring-apache-{context}']
```
Can we add a note that states that users should carefully read the Apache configuration files including all code comments and customize as needed.

kalexand-rh replied:
@enj, line 31.

simo5 commented on install_config/topics/sssd_for_ldap_failover_configure_sssd.adoc:
```diff
@@ -0,0 +1,92 @@
+
+. If you do not want users to log in to the remote basic authentication server by
+using LDAP, edit the *_/etc/pam.d/system-auth_* and
+*_/etc/pam.d/password-auth_*  files and remove the lines that contain *pam_sss.so*.
```
Should probably advice to create proper HBAC rules for IPA, or GPOs for AD.
Removing lines from those files will be reversed during upgrades. It is ok to point to SSSD documentation for futher information instead of providing half-broken guidance here.

stuartchuan commented:
@enj there should be the sssd configuration issue, working on setup the sssd service, but still could not get it workable.

enj commented:
@stuartchuan can you set up an environment and send me the info so I can try to debug it?

kalexand-rh commented on install_config/topics/sssd_for_ldap_failover_configure_sssd.adoc:
```diff
@@ -0,0 +1,94 @@
+using LDAP credentials. You can use either the local console or a remote service
+like SSH to log in.
+
+[IMPORTANT]
```
@simo5, is this better?

mrogers950 commented on install_config/topics/sssd_for_ldap_failover_configure_sssd.adoc:
```diff
@@ -0,0 +1,94 @@
+----
+
+. Grant Apache permission to retrieve the email attribute. Add the following 
+lines of code to the *[ifp]* section of the *_/etc/sssd/sssd.conf_* file:
```
omit "of code"

mrogers950 commented on install_config/topics/sssd_for_ldap_failover_configure_sssd.adoc:
```diff
@@ -0,0 +1,94 @@
+
+* If you use IPA joined systems,
+link:https://www.freeipa.org/page/Howto/HBAC_and_allow_all[configure host-based access control].
+* If you use Active Directory, joined systems, use a
```
"Directory, joined" errant comma

mrogers950 commented on install_config/topics/sssd_for_ldap_failover_configure_apache.adoc:
```diff
@@ -0,0 +1,194 @@
+
+. Enable the module to be loaded by Apache.  To do so, modify the
+*_/etc/httpd/conf.modules.d/55-lookup_identity.conf_* file and uncomment the 
+following line of code:
```
omit "of code"

stuartchuan commented on install_config/topics/sssd_for_ldap_failover_configure_apache.adoc:
```diff
@@ -0,0 +1,194 @@
+    foreach ($env_map as $env_name => $json_name) {
+        $env_data = apache_getenv($env_name)
+        if (!empty($env_data)) {
+            $data[$json_name] = $env_data
```
This line should end with semicolon `；`

stuartchuan commented on install_config/topics/sssd_for_ldap_failover_configure_apache.adoc:
```diff
@@ -0,0 +1,194 @@
+
+    // Add all non-empty environment variables to JSON data
+    foreach ($env_map as $env_name => $json_name) {
+        $env_data = apache_getenv($env_name)
```
This line should end with semicolon `;`

stuartchuan commented on install_config/topics/sssd_for_ldap_failover_configure_openshift.adoc:
```diff
@@ -0,0 +1,66 @@
+user name. Run the following command:
++
+----
+oc get identities -o yaml
```
@enj Seems there no email field in identity.
`# oc get identity
NAME       IDP NAME   IDP USER NAME   USER NAME   USER UID
sssd:pm5   sssd       pm5             pm5         0a6758df-6308-11e8-be63-fa163efb1a65`

enj replied:
@kalexand-rh actually `identities` is probably fine as-is.  We tend to say `pods`, `groups`, etc...

kalexand-rh replied:
`oc get identity` is used more frequently in the collection, so I'll keep the change.

stuartchuan replied:
@enj email still not visible,
`# oc get identity -o yaml
apiVersion: v1
items:
- apiVersion: user.openshift.io/v1
  extra:
    name: pm5 IPA
  kind: Identity
  metadata:
    creationTimestamp: 2018-05-31T07:05:41Z
    name: sssd:pm5
    namespace: ""
    resourceVersion: "6203"
    selfLink: /apis/user.openshift.io/v1/identities/sssd%3Apm5
    uid: 07884bdf-64a1-11e8-a911-fa163e12ab07
  providerName: sssd
  providerUserName: pm5
  user:
    name: pm5
    uid: 0787ffe5-64a1-11e8-a911-fa163e12ab07
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""`

stuartchuan replied:
@enj the email still not visible,
`# oc get identity -o yaml
apiVersion: v1
items:
- apiVersion: user.openshift.io/v1
  extra:
    name: pm5 IPA
  kind: Identity
  metadata:
    creationTimestamp: 2018-05-31T07:05:41Z
    name: sssd:pm5
    namespace: ""
    resourceVersion: "6203"
    selfLink: /apis/user.openshift.io/v1/identities/sssd%3Apm5
    uid: 07884bdf-64a1-11e8-a911-fa163e12ab07
  providerName: sssd
  providerUserName: pm5
  user:
    name: pm5
    uid: 0787ffe5-64a1-11e8-a911-fa163e12ab07
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""`

enj commented on install_config/topics/sssd_for_ldap_failover_configure_apache.adoc:
```diff
@@ -0,0 +1,194 @@
+    
+    # Store attributes in environment variables
+    LookupOutput Env
+    LookupUserAttr mail REMOTE_USER_MAIL
```
@stuartchuan I have checked this manually by `SetEnv REMOTE_USER_MAIL test@test.com` which sets the email in the identity as expected.  Can you confirm that your mod lookup and SSSD set up is correct and that it is sending a non-empty `REMOTE_USER_MAIL` header?  You also need to make sure that `mail` is the attribute that stores the user's email.

I know it is at least partially working because I see this in your YAML:

```yaml
  extra:
    name: pm5 IPA
```

which likely means that either `mail` is the wrong attribute or SSSD is not configured to allow Apache to read it (see `Grant Apache permission to retrieve the email attribute`).

Please also provide the output of `oc get user -o yaml`.

stuartchuan replied:
@enj all the setting following the docs, and how to check it is sending a non-empty REMOTE_USER_MAIL header?
The mail is the attribute that stored in the user:
--------------
1 user matched
--------------
  User login: pm5
  First name: pm5
  Last name: IPA
  Home directory: /home/pm5
  Login shell: /bin/sh
  Principal name: pm5@EXAMPLE.IO
  Principal alias: pm5@EXAMPLE.IO
  Email address: pm5@example.io
  UID: 1655400007
  GID: 1655400007
  Account disabled: False

# oc get user pm5 -o yaml
apiVersion: user.openshift.io/v1
fullName: pm5 IPA
groups: null
identities:
- sssd:pm5
kind: User
metadata:
  creationTimestamp: 2018-06-01T02:41:08Z
  name: pm5
  resourceVersion: "7602"
  selfLink: /apis/user.openshift.io/v1/users/pm5
  uid: 3d3037fa-6545-11e8-989f-fa163e724cb3

# oc get identity sssd:pm5 -o yaml
apiVersion: user.openshift.io/v1
extra:
  name: pm5 IPA
kind: Identity
metadata:
  creationTimestamp: 2018-06-01T02:41:08Z
  name: sssd:pm5
  resourceVersion: "7603"
  selfLink: /apis/user.openshift.io/v1/identities/sssd%3Apm5
  uid: 3d30b1f5-6545-11e8-989f-fa163e724cb3
providerName: sssd
providerUserName: pm5
user:
  name: pm5
  uid: 3d3037fa-6545-11e8-989f-fa163e724cb3

enj replied:
It is unclear what attribute `Email address: pm5@example.io` uses.

I assume the debug logs will show envelope vars?

Reading https://www.adelton.com/apache/mod_lookup_identity may be helpful.

stuartchuan replied:
The attribute is email in the IPA server.
I could find some error hint but still could not fix it.
[Fri Jun 01 04:53:03.600737 2018] [lookup_identity:debug] [pid 8378] mod_lookup_identity.c(445): [client **:47036] invoked for user pm
[Fri Jun 01 04:53:03.610369 2018] [lookup_identity:error] [pid 8378] [client **:47036] dbus call GetUserAttr returned value 0 instead of DBUS_TYPE_DICT_ENTRY

enj replied:
@stuartchuan if `email` is the correct attribute in your setup, then you would have to update your configuration to use `email` instead of `mail` correct?  I find this surprising because IPA docs seem to indicate that the attribute is `mail`.  The error probably indicates that `LookupUserAttr mail REMOTE_USER_MAIL` is failing?

simo5 replied:
@adelton can you help us here ?

adelton replied:
In IPA, the attribute for email address is `mail`.

The return value `0` seems like something is misconfigured on SSSD / SSSD ifp side.

You can verify the operation outside of Apache with

```
dbus-send --print-reply --system --dest=org.freedesktop.sssd.infopipe \
    /org/freedesktop/sssd/infopipe org.freedesktop.sssd.infopipe.GetUserAttr \
    string:pm5 array:string:mail
```

stuartchuan replied:
# dbus-send --print-reply --system --dest=org.freedesktop.sssd.infopipe /org/freedesktop/sssd/infopipe org.freedesktop.sssd.infopipe.GetUserAttr string:pm5 array:string:mail
method return time=1528093813.798625 sender=:1.2787 -> destination=:1.2810 serial=23 reply_serial=2
   array [
   ]

When string setting to user email could get results,
# dbus-send --print-reply --system --dest=org.freedesktop.sssd.infopipe /org/freedesktop/sssd/infopipe org.freedesktop.sssd.infopipe.GetUserAttr string:pm5@example.io array:string:mail
method return time=1528091855.672691 sender=:1.2787 -> destination=:1.2795 serial=13 reply_serial=2
   array [
      dict entry(
         string "mail"
         variant             array [
               string "pm5@example.io"
            ]
      )
   ]

stuartchuan replied:
When login by user email to the cluster, the mail field will present to the identity field.
```
# oc get user -o  yaml
apiVersion: v1
items:
- apiVersion: user.openshift.io/v1
  fullName: pm5 IPA
  groups: null
  identities:
  - sssd:pm5@example.io
  kind: User
  metadata:
    creationTimestamp: 2018-06-04T06:25:33Z
    name: pm5@example.io
    namespace: ""
    resourceVersion: "553293"
    selfLink: /apis/user.openshift.io/v1/users/pm5%40example.io
    uid: 15f75ffe-67c0-11e8-989f-fa163e724cb3
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""

# oc get identity -o yaml
apiVersion: v1
items:
- apiVersion: user.openshift.io/v1
  extra:
    email: pm5@example.io
    name: pm5 IPA
  kind: Identity
  metadata:
    creationTimestamp: 2018-06-04T06:25:33Z
    name: sssd:pm5@example.io
    namespace: ""
    resourceVersion: "553294"
    selfLink: /apis/user.openshift.io/v1/identities/sssd%3Apm5%40example.io
    uid: 15f7a78e-67c0-11e8-989f-fa163e724cb3
  providerName: sssd
  providerUserName: pm5@example.io
  user:
    name: pm5@example.io
    uid: 15f75ffe-67c0-11e8-989f-fa163e724cb3
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

adelton replied:
Not sure I understand but it looks like `pm5` is then not the identity SSSD would be able to resolve, and the fully qualified id has to be used.

What does `id pm5` and `id pm5@example.io` on the command line say?

stuartchuan replied:
Should be the same user.
```
# id pm5
uid=1655400007(pm5) gid=1655400007(pm5@example.io) groups=1655400007(pm5@example.io)
# id pm5@example.io
uid=1655400007(pm5@example.io) gid=1655400007(pm5@example.io) groups=1655400007(pm5@example.io)
```

adelton replied:
In that case, it might be worth increasing the debug level in SSSD (or at least for its `ifp`) and check why over D-Bus, there is no reasonable reply for the short identifier (`pm5`).

stuartchuan replied:
Here is the debug level 10 for ifp, is there any hint?
```
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_dispatch] (0x4000): dbus conn: 0x7f4a0e9bfe40
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_dispatch] (0x4000): Dispatching.
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_signal_handler] (0x2000): Received D-Bus signal org.freedesktop.DBus.NameOwnerChanged
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_signal_handler_got_caller_id] (0x0400): Got a signal from the bus..
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_signal_name_owner_changed] (0x0400): Clearing UIDs cache
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_dispatch] (0x4000): dbus conn: 0x7f4a0e9bfe40
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_dispatch] (0x4000): Dispatching.
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_message_handler] (0x2000): Received SBUS method org.freedesktop.sssd.infopipe.GetUserAttr on path /org/freedesktop/sssd/infopipe
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_get_sender_id_send] (0x2000): Looking for identity of sender [:1.2849]
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_add_timeout] (0x2000): 0x7f4a0e9ba5b0
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_remove_timeout] (0x2000): 0x7f4a0e9ba5b0
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_dispatch] (0x4000): dbus conn: 0x7f4a0e9bfe40
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_dispatch] (0x4000): Dispatching.
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ifp_user_get_attr] (0x0040): GetUserAttr is deprecated, please consider switching to org.freedesktop.sssd.infopipe.Users.User interface
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ifp_user_get_attr] (0x0200): Looking up attributes of user [pm5] on behalf of 0
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sss_parse_name_for_domains] (0x0200): name 'pm5' matched without domain, user is pm5
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_set_plugin] (0x2000): CR #5: Setting "User by name" plugin
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_send] (0x0400): CR #5: New request 'User by name'
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_process_input] (0x0400): CR #5: Parsing input name [pm5]
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sss_parse_name_for_domains] (0x0200): name 'pm5' matched without domain, user is pm5
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_set_name] (0x0400): CR #5: Setting name [pm5]
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_select_domains] (0x0400): CR #5: Performing a multi-domain search
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_search_domains] (0x0400): CR #5: Search will check the cache and check the data provider
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_validate_domain_type] (0x2000): Request type Any for domain example.io type POSIX is valid
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_set_domain] (0x0400): CR #5: Using domain [example.io]
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_prepare_domain_data] (0x0400): CR #5: Preparing input data for domain [example.io] rules
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_search_send] (0x0400): CR #5: Looking up pm5@example.io
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_search_ncache] (0x0400): CR #5: Checking negative cache for [pm5@example.io]
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sss_ncache_check_str] (0x2000): Checking negative cache for [NCE/USER/example.io/pm5@example.io]
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_search_ncache] (0x0400): CR #5: [pm5@example.io] is not present in negative cache
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_search_cache] (0x0400): CR #5: Looking up [pm5@example.io] in cache
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Added timed event "ltdb_callback": 0x7f4a0e9d1090

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Added timed event "ltdb_timeout": 0x7f4a0e9d2c90

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Running timer event 0x7f4a0e9d1090 "ltdb_callback"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Destroying timer event 0x7f4a0e9d2c90 "ltdb_timeout"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Ending timer event 0x7f4a0e9d1090 "ltdb_callback"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Added timed event "ltdb_callback": 0x7f4a0e9cea20

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Added timed event "ltdb_timeout": 0x7f4a0e9d1090

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Running timer event 0x7f4a0e9cea20 "ltdb_callback"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Destroying timer event 0x7f4a0e9d1090 "ltdb_timeout"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Ending timer event 0x7f4a0e9cea20 "ltdb_callback"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_search_send] (0x0400): CR #5: Returning [pm5@example.io] from cache
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_search_ncache_filter] (0x0400): CR #5: This request type does not support filtering result by negative cache
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_create_and_add_result] (0x0400): CR #5: Found 1 entries in domain example.io
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [cache_req_done] (0x0400): CR #5: Finished: Success
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Added timed event "ltdb_callback": 0x7f4a0e9cea20

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Added timed event "ltdb_timeout": 0x7f4a0e9d3ec0

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Running timer event 0x7f4a0e9cea20 "ltdb_callback"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Destroying timer event 0x7f4a0e9d3ec0 "ltdb_timeout"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Ending timer event 0x7f4a0e9cea20 "ltdb_callback"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Added timed event "ltdb_callback": 0x7f4a0e9cea20

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Added timed event "ltdb_timeout": 0x7f4a0e9c36d0

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Running timer event 0x7f4a0e9cea20 "ltdb_callback"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Destroying timer event 0x7f4a0e9c36d0 "ltdb_timeout"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ldb] (0x4000): Ending timer event 0x7f4a0e9cea20 "ltdb_callback"

(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [ifp_user_get_attr_handle_reply] (0x0080): Attribute mail not present or has no values
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_dispatch] (0x4000): dbus conn: 0x7f4a0e9bfe40
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_dispatch] (0x4000): Dispatching.
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_signal_handler] (0x2000): Received D-Bus signal org.freedesktop.DBus.NameOwnerChanged
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_signal_handler_got_caller_id] (0x0400): Got a signal from the bus..
(Mon Jun  4 04:26:36 2018) [sssd[ifp]] [sbus_signal_name_owner_changed] (0x0400): Clearing UIDs cache
```

adelton replied:
And what's the debug output for the fully qualified username?

stuartchuan replied:
The dubug output for user `pm5`

adelton replied:
I understand that. And we'd like to see the debug output when the `pm5@example.io` is used, so see where things start to differ.

adelton replied:
Also, please show the debug logs for the domain, for both cases.

stuartchuan replied:
Here is the sssd.config
```
[domain/example.io]
cache_credentials = True
krb5_store_password_if_offline = True
ipa_domain = example.io
id_provider = ipa
auth_provider = ipa
access_provider = ipa
ldap_tls_cacert = /etc/ipa/ca.crt
ipa_hostname = www.example.io
chpass_provider = ipa
dyndns_update = True
ipa_server = www.example.io
dyndns_iface = ens3
dns_discovery_domain = example.io
realmd_tags = manages-system
default_shell = /bin/bash
use_fully_qualified_names = True
fallback_homedir = /home/%u@%d
ldap_user_extra_attrs = mail
debug_level = 6
[domain/default]
ldap_user_extra_attrs = mail
autofs_provider = ldap
ldap_schema = rfc2307bis
krb5_kpasswd = www.example.io:749
ldap_search_base = cn=users,cn=compat,dc=example,dc=io
krb5_server = www.example.io:88
id_provider = ldap
auth_provider = krb5
chpass_provider = krb5
krb5_store_password_if_offline = True
ldap_uri = ldap://www.example.io/
krb5_realm = EXAMPLE.IO
ldap_id_use_start_tls = True
cache_credentials = True
#ldap_tls_cacertdir = /etc/openldap/cacerts
ldap_tls_cacertdir = /etc/ipa
debug_level = 6
[sssd]
services = nss, sudo, pam, autofs, ssh, ifp, pac

domains = default, example.io
[nss]
homedir_substring = /home

[pam]

[sudo]

[autofs]

[ssh]

[pac]

[ifp]
user_attributes =  +mail
allowed_uids = apache, root
debug_level = 10
[secrets]

[session_recording]
```
The problem should be in the sssd configuration, there were two domains in the configuration, when login by `pm5` will direct to the default domain, and login by `pm5@example.io` will direct to example.io domain.
When I removed default domain from the configuration, only the `pm5@example.io` could login, and with mail in the identity. But when I removed example.io and left the defaut domain only, and change the default to example.io, both `pm5` and `pm5@example.io` could login, but no mail in the identity.

Would you point me how to modify the configuration file to let it working normally?

enj replied:
Remove the `use_fully_qualified_names` from example domain and remove the default domain section completely?

simo5 replied:
Definitely remove the double domains, you want only one.
About the use of fully qualified option or not, that is up to the deployment to decide but we can give guidance for people if they want to use short names.

stuartchuan replied:
@enj suggestion was workable.
@adelton @simo5 @enj thank you all for the support.
No comments to the docs now, and all the steps in the guide were verified.

enj commented on install_config/topics/sssd_for_ldap_failover_configure_sssd.adoc:
```diff
@@ -51,8 +51,40 @@ and add this attribute:
+      )
+   ]
+----
+<1> Provide an email address that your LDAP solution uses.
```
This is the username of a user, so it does not have to be an email.

enj commented on install_config/topics/sssd_for_ldap_failover_configure_sssd.adoc:
```diff
@@ -51,8 +51,40 @@ and add this attribute:
+your domain:
++
+----
+# dbus-send --print-reply --system --dest=org.freedesktop.sssd.infopipe \
```
This probably will not work until after you restart SSSD, which is a later step.

enj commented on install_config/topics/sssd_for_ldap_failover_configure_sssd.adoc:
```diff
@@ -51,8 +51,40 @@ and add this attribute:
+solution. For IPA, specify `mail`. Other LDAP solutions might use another
+attribute, such as `email`.
+
+. Modify the *_domain_* parameter in the *_/etc/sssd/sssd.conf_* file to contain
```
I would say more along the lines of "confirm that it has only one domain"

enj commented on install_config/topics/sssd_for_ldap_failover_configure_sssd.adoc:
```diff
@@ -51,8 +51,40 @@ and add this attribute:
+----
+# dbus-send --print-reply --system --dest=org.freedesktop.sssd.infopipe \
+    /org/freedesktop/sssd/infopipe org.freedesktop.sssd.infopipe.GetUserAttr \
+    string:pm5@example.io \ <1>
```
`string:username`

enj commented on install_config/topics/sssd_for_ldap_failover_configure_sssd.adoc:
```diff
@@ -51,8 +51,40 @@ and add this attribute:
+      dict entry(
+         string "mail"
+         variant             array [
+               string "pm5@example.io"
```
`string "username@example.com"`

enj commented on install_config/topics/sssd_for_ldap_failover_configure_sssd.adoc:
```diff
@@ -108,6 +86,28 @@ $ getent passwd <username>
+      dict entry(
+         string "mail"
+         variant             array [
+               string "username"@yourdomainname.com"
```
Quotation mark out of place.

enj commented on install_config/topics/sssd_for_ldap_failover_configure_sssd.adoc:
```diff
@@ -57,35 +57,13 @@ ldap_user_extra_attrs = mail <1>
-. Modify the *_domain_* parameter in the *_/etc/sssd/sssd.conf_* file to contain
-only one domain name.
+. Confirm that the *_domain_* parameter in the *_/etc/sssd/sssd.conf_* file
+contains only one domain name.
```
If you think it would be helpful, something like

`... contains only one domain name and a single macthing section [domain/yourdomainname.com]`

If you do not think it would be helpful, the current wording is fine.

kalexand-rh commented:
@openshift/team-documentation, PTAL. QE and engineering have signed off on the change, and a refreshed built version is here: http://file.rdu.redhat.com/kalexand/052918/pr9025/install_config/sssd_for_ldap_failover.html

vikram-redhat commented on install_config/topics/sssd_for_ldap_failover_overview.adoc:
```diff
@@ -0,0 +1,31 @@
+
+[IMPORTANT]
+====
+You must complete all sections of this topic.
```
@kalexand-rh - Hey Kathryn, you may need to expand this note. In single page views on the portal, this may cause confusion (multi-page and d.o.c it is clear what it means).

ahardin-rh commented on install_config/topics/sssd_for_ldap_failover_configure_openshift.adoc:
```diff
@@ -0,0 +1,66 @@
+    --node-config=/etc/origin/node-node1.example.com/node-config.yaml
+----
+
+. Test a login by using the `oc` cli:
```
s/cli/CLI

ahardin-rh commented on install_config/topics/sssd_for_ldap_failover_configure_openshift.adoc:
```diff
@@ -0,0 +1,66 @@
+user name. Run the following command:
++
+----
+oc get identity -o yaml
```
Missing the $ before the command

ahardin-rh commented:
@kalexand-rh Just a few nits from me. Other than that, looks good. :star2:

kalexand-rh commented:
Thank you all. :) Merging.

kalexand-rh commented:
/cherrypick enterprise-3.10

kalexand-rh commented:
/cherrypick enterprise-3.9

kalexand-rh commented:
/cherrypick enterprise-3.7

kalexand-rh commented:
/cherrypick enterprise-3.6

kalexand-rh commented:
/cherrypick enterprise-3.5

kalexand-rh commented:
/cherrypick enterprise-3.4

kalexand-rh commented:
/cherrypick enterprise-3.3

kalexand-rh commented:
Oh, wow, I don't think I've seen them all fail like that.
3.10: https://github.com/openshift/openshift-docs/pull/9999
3.9: https://github.com/openshift/openshift-docs/pull/10000
3.7: https://github.com/openshift/openshift-docs/pull/10001

The rest of the cherry-picks are messy, and I'll fix them in the morning.

kalexand-rh commented:
3.6: https://github.com/openshift/openshift-docs/pull/10002
3.5: https://github.com/openshift/openshift-docs/pull/10003
3.4: https://github.com/openshift/openshift-docs/pull/10004
3.3: https://github.com/openshift/openshift-docs/pull/10005

enj commented:
The link in this section is broken, has extra .adoc in URL.

https://docs.openshift.org/latest/install_config/configuring_authentication.html#BasicAuthPasswordIdentityProvider

vikram-redhat commented:
Hey @kalexand-rh - something went wrong in the merges for 3.6 to 3.3 branches. They are all missing the changes for the _topic_map.yml (for one thing), but also to other files, like manage_users.adoc (3.6-3.4) and expose_internal_services.adoc (+2 more)(3.3). I can understand files other than _topic_map.yml being not there for older branches, but the changes for _topic_map.yml should be there in all. Curiously, 3.10-3.7 are all correct.

kalexand-rh commented:
- _topic_map.yml - I misinterpreted what git was doing with this conflict. Fixed in the PR.
- sssd_for_ldap_failover.adoc - this file should be removed. Fixed in the PR.
- admin_guide/manage_users.adoc - The changed section isn't in 3.6 or earlier
- dev_guide/expose_service/expose_internal_ip_service.adoc - This topic isn't in 3.3.
- security/securing_container_platform.adoc - This topic isn't in 3.4 or earlier.

https://github.com/openshift/openshift-docs/pull/10040
